### ISSUE TITLE
Generation failing for enum with version >= 2.11.472

### ISSUE THREAD

**Issue by antch** [vojtechhabarta/typescript-generator#327]
I had no issue generating types for my entire application until 2.11.472, where I suddenly get this stack:

```
java.lang.ExceptionInInitializerError
	at com.mypackage.common.MyEnum.<clinit>(MyEnum.java:26)
	at sun.misc.Unsafe.ensureClassInitialized(Native Method)
	at sun.reflect.UnsafeFieldAccessorFactory.newFieldAccessor(UnsafeFieldAccessorFactory.java:43)
	at sun.reflect.ReflectionFactory.newFieldAccessor(ReflectionFactory.java:156)
	at java.lang.reflect.Field.acquireFieldAccessor(Field.java:1088)
	at java.lang.reflect.Field.getFieldAccessor(Field.java:1069)
	at java.lang.reflect.Field.get(Field.java:393)
	at cz.habarta.typescript.generator.parser.Jackson2Parser.parseEnumOrObjectEnum(Jackson2Parser.java:467)
	at cz.habarta.typescript.generator.parser.Jackson2Parser.parseClass(Jackson2Parser.java:181)
	at cz.habarta.typescript.generator.parser.ModelParser.parseQueue(ModelParser.java:88)
	at cz.habarta.typescript.generator.parser.ModelParser.parseModel(ModelParser.java:50)
	at cz.habarta.typescript.generator.TypeScriptGenerator.generateTypeScript(TypeScriptGenerator.java:66)
	at cz.habarta.typescript.generator.TypeScriptGenerator.generateTypeScript(TypeScriptGenerator.java:57)
	...<snip>...
Caused by: java.util.NoSuchElementException
	at java.util.ServiceLoader$LazyIterator.nextService(ServiceLoader.java:365)
	at java.util.ServiceLoader$LazyIterator.next(ServiceLoader.java:404)
	at java.util.ServiceLoader$1.next(ServiceLoader.java:480)
	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:307)
	...<snip>...
	at com.mypackage.common.MyEnum$Constants.<clinit>(MyEnum.java:150)
	... 32 more
```

Now it appears to be actually trying to instantiate the enums, which is causing an error as the code underneath is performing operations that look to depend on other aspects of the 
application's environment.

In my case the enum constructor takes a parameter which comes from a `static final` variable, like:

```java
public enum MyEnum {

    MY_VALUE(Constants.OPTIONS);

    private static class Constants {
        static final ValueDefinition OPTIONS = MyUtil.doesSomethingNonTrivial(SomeOtherClass.class);
    }
}
```

Is there any way to use the old behavior? This enum generates with no issues on 2.10 or earlier.

**Comment by vojtechhabarta** [comment]
I identified commit which changed the behavior: e27accad6ec978ddc4dfe5447f9945a5a24b1206.

This commit handles Jackson feature that uses `toString()` method for determining enum values. It also made typescript-generator more consistent with Jackson (while removing chunk of custom code handling enums).

What I can do is to catch the error and fallback to name of the enum constant.

**Comment by vojtechhabarta** [comment]
I committed mentioned improvement.

**Comment by antch** [comment]
Hey thank you, this works!

**Comment by antch** [comment]
Do you think it would be reasonable to add the same fallback when the value is not a number or string, instead of just ignoring the enum constant? My situation is that I have a bunch of (63, to be exact) enums that implement an interface that has its own custom serializer/deserializer. The custom serializer produces a Map, so every enum constant gets skipped and I end up with an empty type. Or, maybe there's a different way around the issue?

**Comment by vojtechhabarta** [comment]
Typescript-generator's goal is to describe JSON data produced from Java classes. So if your custom serializer produces Map it should not be described as enum.
Or what's the reason to have TypeScript declarations that doesn't describe real data?

**Comment by antch** [comment]
So basically this is the structure:

```
public interface CodedValue {
  String code;
  String definition;
}

public enum MyEnum implements CodedValue {
  MY_FIRST_VALUE("A0", "Some description");
}

public interface MyClassUsingEnum {
  public MyEnum getSomeValue();
}
```

I agree with you that the constant part isn't really very important, but I don't see a way with typescript-generator to easily achieve a set of logical type definitions for the above. I may be missing something, though. I will say that having the constant is better than nothing at all IMO.

**Comment by vojtechhabarta** [comment]
I would split this case into two parts.

First thing is to have right TypeScript types. In general it is not possible to automatically determine correct static types when custom serializers are used. But typescript-generator has `customTypeMappings` parameter which can be used to manually change type. So in this case you can map `MyEnum` to general map/object type or even to `{ code: string, definition: string }` type. I committed unit test with this case: add6b4c0a16d68afce6787b3197d98314a1d392a.

Second thing is delivering data like `MY_FIRST_VALUE("A0", "Some description")` to the client. Instead of TypeScript file I would use JSON which I would expose using REST API or generate in build time.

**Comment by antch** [comment]
Thanks for that test. I think that would do what I need, the only issue is that I have over 60 of these enums which is going to make for a very busy (and difficult to maintain) configuration of the generator.

**Comment by vojtechhabarta** [comment]
Yes, using custom `TypeProcessor` should work well, you just need to add your compiled class to plugin dependencies.

**Comment by vojtechhabarta** [comment]
Commit "Catching error thrown when determining enum value" is now released in [2.14.505](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.14.505).
